rootfs/etc/nginx/lua/balancer.lua

```diff
@@ -122,28 +98,46 @@ function _M.init_worker()
   end
 end
 
-function _M.call()
-  local phase = ngx.get_phase()
-  if phase == "log" then
-    after_balance()
-    return
+function _M.rewrite()
+  local balancer = get_balancer()
+  if not balancer then
+    ngx.status = ngx.HTTP_SERVICE_UNAVAILABLE
+    return ngx.exit(ngx.status)
   end
-  if phase ~= "balancer" then
-    return error("must be called in balancer or log, but was called in: " .. phase)
+end
+
+function _M.balance()
+  local balancer = get_balancer()
+  if not balancer then
+    return
   end
 
-  local host, port = balance()
+  local host, port = balancer:balance()
   if not host then
-    ngx.status = ngx.HTTP_SERVICE_UNAVAILABLE
-    return ngx.exit(ngx.status)
+    ngx.log(ngx.WARN, "no host returned, balancer: " .. balancer.name)
+    return
   end
 
   ngx_balancer.set_more_tries(1)
 
   local ok, err = ngx_balancer.set_current_peer(host, port)
   if not ok then
-    ngx.log(ngx.ERR, string.format("error while setting current upstream peer to %s", tostring(err)))
+    ngx.log(ngx.ERR, "error while setting current upstream peer to " .. tostring(err))
   end
 end
 
+function _M.log()
+  local balancer = get_balancer()
+  if not balancer then
+    return
+  end
+
+  balancer:after_balance()
```

Comment:
I think this should be checking if the function exists, since it can be optional.